core/sail/memory/src/main/java/org/eclipse/rdf4j/sail/memory/model/MemStatementList.java

```diff
@@ -75,69 +72,73 @@ public void add(MemStatement st) throws InterruptedException {
 			}
 		}
 
-		long readLock = addRemoveLock.readLock();
-		try {
-			do {
-
-				MemStatement[] statements = getStatements();
-				int length = statements.length;
+		do {
 
-				if (length > (int) SIZE.getAcquire(this)) {
+			MemStatement[] statements = getStatements();
+			int length = statements.length;
 
-					int previouslyInsertedIndex = (int) PREVIOUSLY_INSERTED_INDEX.getOpaque(this);
-					if (previouslyInsertedIndex >= length) {
-						continue;
-					}
+			boolean shouldGrowArray = true;
 
-					int i = previouslyInsertedIndex + 1 >= length ? 0 : previouslyInsertedIndex + 1;
+			if (length > (int) SIZE.getAcquire(this)) {
 
-					for (; i != previouslyInsertedIndex; i = (i + 1 >= length ? 0 : i + 1)) {
+				int previouslyInsertedIndex = (int) PREVIOUSLY_INSERTED_INDEX.getOpaque(this);
+				if (previouslyInsertedIndex >= length) {
+					continue;
+				}
 
-						if (statements[i] == null) {
+				int i = previouslyInsertedIndex + 1 >= length ? 0 : previouslyInsertedIndex + 1;
 
-							boolean success = STATEMENTS_ARRAY.compareAndSet(statements, i, null, st);
-							if (success) {
+				for (; i != previouslyInsertedIndex; i = (i + 1 >= length ? 0 : i + 1)) {
 
-								// check if the statements array has been swapped out (because it was grown) while we
-								// were
-								// inserting into it
-								MemStatement[] statementsAfterInsert = getStatements();
-								if (statementsAfterInsert != statements
-										&& STATEMENTS_ARRAY.getAcquire(statements, i) != st) {
-									// we wrote into an array while it was growing and our write was lost
-									break;
-								}
```

Comment:
This is actually a bug that could lead to a statement not getting inserted if there is a highly concurrent write load.